File: make.pl
#!/bin/perl

use strict;
use warnings;

my $CC = 'gcc';
my $FLAGS = '-g -O0';
my $SOURCES = 'main.c common.c diagnostic.c lexer.c parser.c semantic.c ir.c';
my $TARGET = 'bin/seal';

sub change_test 
{
    my ($i) = @_;
    my $dosya = 'main.c';

    open my $in, '<', $dosya or die $!;
    my @lines = <$in>;
    close $in;

    if ($i eq "lexer") 
    {
        for (@lines) 
        {
            s/run_parser\(\)/\/\/run_parser()/;
            s/print_ast\(1\)/\/\/print_ast(1)/;
            s/\/\/run_lexer_machine\(sources_list\)/run_lexer_machine(sources_list)/;
            s/\/\/print_tokens\(1\)/print_tokens(1)/;
        }
    }
    
    if ($i eq "parser") 
    {
        for (@lines) 
        {
            s/\/\/run_parser\(\)/run_parser()/;
            s/\/\/print_ast\(1\)/print_ast(1)/;
            s/print_tokens\(1\)/\/\/print_tokens(1)/;
        }
    }

	if ($i eq "close")
	{
		s/print_ast\(1\)/\/\/print_ast(1)/;
		s/print_tokens\(1\)/\/\/print_tokens(1)/;
	}
	
    if ($i eq "open_all")
    {
    	for (@lines)
    	{
    		s/\/\/run_parser\(\)/run_parser()/;
    		s/\/\/print_ast\(1\)/print_ast(1)/;
            s/\/\/run_lexer_machine\(sources_list\)/run_lexer_machine(sources_list)/;
            s/\/\/print_tokens\(1\)/print_tokens(1)/;
    	}
    }

    open my $out, '>', $dosya or die $!;
    print $out @lines;
    close $out;
}

sub print_out
{
	print "MAKESCRIPT: OUT -> ";
	system("echo $?");
}

if (scalar @ARGV < 1)
{
	my $command = "$CC $FLAGS $SOURCES -o $TARGET";

	system($command) == 0 or die "\nbuild error";
	print "MAKESCRIPT: $command\n";
	print_out();
	exit 0;
}

if ($ARGV[0] eq "clear")
{
	my $command = "rm $TARGET";
	
	system($command) == 0 or die "\n$TARGET not exists";
	print "MAKESCRIPT: $command\n";
	print_out();
	exit 0;
}

if ($ARGV[0] eq "run" || $ARGV[0] eq "mrun")
{
	my $command = "./bin/seal $ARGV[1]";

	if ($ARGV[0] eq "mrun")
	{
		$command = "$CC $FLAGS $SOURCES -o $TARGET && ./bin/seal $ARGV[1]";
	}
	
	system($command) == 0 or die "\n$TARGET not exists";
	print "MAKESCRIPT: $command\n";
	print_out();
	exit 0;
}

if ($ARGV[0] eq "test")
{
	change_test($ARGV[1]);
	print("Changed $ARGV[1]");
	exit 0;
}

print "Wrong arg\n";
exit 1;

File: make.lua
#!/bin/lua

cc = "gcc";
flags = "-g -O0";
sources = "main.c common.c preprocessor/preprocessor.c diagnostic.c lexer.c parser.c semantic.c ir.c codegen.c";
target = "bin/seal";

function make()
    local success, exit_type, code = os.execute(cc .. " " .. flags .. " " .. sources .. " -o " .. target);
    print("OUTPUT: " .. target .. "\n");
    print("SUCCESS: " .. tostring(success));
    print("EXIT TYPE: " .. tostring(exit_type));
    print("CODE: " .. tostring(code) .. "\n");
    
    print(cc .. " " .. flags .. " " .. sources .. " -o " .. target);
    os.exit(0);
end

function install()
    os.execute("mv " .. target .. arg[2]);
end

function run()
    os.execute("./" .. target .. " --Save llvm --Compile " .. arg[2]);
end

function clear()
    os.execute("rm " .. target);
end

CMD = arg[1];

if CMD == "install" or CMD == "run" then
    sec_arg = arg[2];

    if sec_arg == " " or nil then
        print("Missing secondary Argument.\n");
        os.exit(1);
    end
end

if arg[1] == nil or arg[1] == " " then
    make();
elseif CMD == "install" then
    install();
elseif CMD == "run" then
    run();
elseif CMD == "clear" then
    clear();
else
    print("Missing primary argument\n");
end
